Introduce MetadataValidators to test arbitrary metadata for freshness.

bnw opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #7230 |
| Related ticket | #7176 |
| License | MIT |
| Doc PR | TODO |

This PR implements the idea discussed here: https://github.com/symfony/AsseticBundle/issues/168#issuecomment-13608074
## Idea

Currently, any metadata that is passed to ConfigCache (namely implementations of ResourceInterface) is serialized to disk. Upon freshness check, it is unserialized and then asked, whether it is fresh.
The metadata has no chance to interact with any services, allowing only for very basic freshness checks.

With this PR, metadata no longer needs to implement the check for freshness itself. Instead, the ConfigCache::isFresh gets handed an array of MetadataValidator that take over the job of checking for freshness.
Each validator supports certain types of metadata. The implementations of ResourceInterface are now only a special type of metadata, that is validated using the ResourceInterfaceValidator. 

In most cases, the ConfigCache should be used via the ConfigCacheFactory. I added a config_cache_factory service to which one can add validators via tagged services.
## BC

Calling ConfigCache::isFresh without arguments is now deprecated.
Using ConfigCacheFactory without adding validators is deprecated.
However, both are still supported and yield the same results as before.

bnw committed:
Introduce MetadataValidators to test arbitrary metadata for freshness.

bnw committed:
Added trailing newline

bnw committed:
Added new parameter to interface

bnw committed:
Renamed ResourceValidator to ResourceInterfaceValidator
Added test for ResourceInterfaceValidator
ConfigCache::isFresh: Stop validation as soon as we found a supporting validator

bnw committed:
Added missing import

bnw committed:
Correcting classname after renaming

bnw committed:
Added ConfigCachePass to framework

bnw committed:
Fixed UnitTests

bnw committed:
Bugfix

bnw committed:
Better name

iltar commented on src/Symfony/Component/Config/MetadataValidatorInterface.php:
```diff
@@ -0,0 +1,26 @@
+    public function supports($metadata);
+
+    public function isFresh($metadata, $timestamp);
+
```
Missing docs for those 2 methods

iltar commented on src/Symfony/Component/Config/Resource/ResourceInterfaceValidator.php:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @author Benjamin Klotz <bk@webfactory.de>
+ */
+class ResourceInterfaceValidator implements MetadataValidatorInterface
```
What about just calling this `ResourceValidator`?

ogizanagi replied:
+1

mpdude replied:
Renamed it.

mpdude committed:
Fix test

mpdude committed:
Add docblock

mpdude committed:
Fix tests while keeping ConfigCacheFactory unchanged (from a client perspective)

mpdude committed:
Hello Fabbot

mpdude committed:
Forgot import statement

mpdude committed:
The new ConfigCachePass is required

mpdude committed:
Bump deps

ogizanagi commented on src/Symfony/Component/Config/ValidatorConfigCacheFactory.php:
```diff
@@ -0,0 +1,64 @@
+ *
+ * @author Matthias Pigulla <mp@webfactory.de>
+ */
+class ValidatorConfigCacheFactory implements ConfigCacheFactoryInterface
```
Why creating a new config cache factory given adding validators to the old one is doable without BC break ? (But will prevent the deprecation due to https://github.com/symfony/symfony/pull/15692/files#diff-9f40590069e1fa6582e56bcea49e09a4R77)
Will the `ConfigCacheFactory` trigger a deprecation saying to use the new `ValidatorConfigCacheFactory` instead ?

mpdude replied:
For BC reasons, the `Router` and `Translator` need to create a `ConfigCacheFactory` themselves if none is injected. If that factory comes without the default `ResourceInterfaceValidator`, they'd have to add one themselves and that would cause unnecessary dependencies. Having a default `ResourceInterfaceValidator` in `ConfigCacheFactory` would be weird as well.

So this way, use a `ConfigCacheFactory` if you don't care about special validations and you'll get the same behaviour as in the past.

Use a ValidatorConfigCacheFactory if you care about special validations or if you get the instance injected anyway and don't care about the hassle of setting it up.

ogizanagi commented on src/Symfony/Component/Config/DependencyInjection/ConfigCachePass.php:
```diff
@@ -0,0 +1,33 @@
+        $definition = $container->getDefinition('config_cache_factory');
+
+        foreach ($container->findTaggedServiceIds('config_cache.metadata_validator') as $id => $attributes) {
+            $definition->addMethodCall('addValidator', array(new Reference($id)));
```
Would it make sense to handle priorities at this point ?

mpdude replied:
I think that there should not be more than one authoritative `MetadataValidator` per resource type and if there are more, they should not disagree about freshness. 

Also, in order to keep things simple, we decided against coming up with a full-fledged voting strategy like in the Security component.

mpdude committed:
Avoid changing the interface

mpdude committed:
Also validate this method in the interface

mpdude committed:
Rename ResourceInterfaceValidator to ResourceValidator as suggested on GH

mpdude committed:
Oh Fabbot

mpdude committed:
Add a note why the ResourceInterface::__toString method is important and unlikely to disappear in the future

mpdude committed:
Remove totally broken test

This may have worked in the past, but was based on assumptions how the cache was implemented. You cannot dump a cache in prod and then expect it to be fresh or not when switching back to dev.

A useful test would:

1. Use a mock loader to detect reloads
2. In prod: Use a Translator to prime the cache; try again with another instance and see that the cache is used
3. In dev: Use a Translator to prime the cache; add an "always stale" resource; try again with another instance and see that the cache is refreshed.

mpdude committed:
Rewrite ConfigCacheTest because it made many assumptions about how the cache works internally

mpdude committed:
Fix ConfigCache and improve docblock

mpdude committed:
If I had to pay just $0.02 for every Fabbot complaint, Fabien would be a rich man already

mpdude commented:
Travis fails in deps=high only due to broken test on the 2.8 branch.

mpdude committed:
Document deprecation in the CHANGELOG file

mpdude committed:
Some more tweaks to the ConfigCacheTest

mpdude committed:
Remove unused import

mpdude commented:
Failed test is Symfony\Component\Console\Tests\Helper\ProgressBarTest::testAnsiColorsAndEmojis on HHVM; seems unrelated.

stof commented on src/Symfony/Component/Config/ConfigCache.php:
```diff
@@ -69,28 +70,61 @@ public function getPath()
+     * The first MetadataValidator that supports a given resource is considered authoritative.
+     * Resources with no matching MetadataValidators will silently be ignored and considered fresh.
+     *
+     * This method <em>does not</em> take the debug flag into consideration: Whether or not a cache
```
this looks weird to me. We need to take the debug flag into account

mpdude replied:
I'd like to make that a decision of the client (the ConfigCacheFactory). In production, it may skip all freshness checks or use a special set of Validators only.

Then, the cache itself would not need to know about `debug` at all.

stof commented on src/Symfony/Component/Config/ConfigCache.php:
```diff
@@ -69,28 +70,61 @@ public function getPath()
+     *
+     * @return bool True if all supported resources and valid, false otherwise
+     */
+    public function isValid(array $validators = null)
```
passing validators to ``isValid`` looks weird to me. Validators should be passed in the constructor IMO (which means we can also keep ``isFresh`` btw)

mpdude replied:
But `isValid` has slightly different semantics, see above.

mpdude replied:
I was able to move the validators to the constructor of a new ValidatorConfigCache class. That way, we can keep the interface without deprecations. 

ConfigCache is a subclass and behaves as previously, switching between dev and prod mode and validating ResourceInterface instances only.

stof commented on src/Symfony/Component/Config/ConfigCache.php:
```diff
@@ -69,28 +70,61 @@ public function getPath()
+                    continue; // next validator
+                }
+                if ($validator->isFresh($resource, $time)) {
+                    continue; // no need to further check this resource
```
this should be a break, not a continue. Otherwise you are checking the following validators on this resource

mpdude replied:
Good catch!

mpdude committed:
Continue with next resource, not next validator

mpdude committed:
Add a shortcut for when we don't have any validators (might be in prod)

mpdude committed:
Create a new ValidatorConfigCache class and pass metadata validators in the constructor.

That way, we can keep the isFresh() method and get away without deprecations.

mpdude committed:
Fabbot

mpdude committed:
This is not relevant for this PR

stof commented on src/Symfony/Component/Config/MetadataValidatorInterface.php:
```diff
@@ -0,0 +1,46 @@
+ *
+ * @author Benjamin Klotz <bk@webfactory.de>
+ */
+interface MetadataValidatorInterface
```
I'm not sure MetadataValidator is the best name. This is more a resource checker than a validator

mpdude replied:
I'm open for naming suggestions. 

* `CacheChecker`
* `CacheMetadataChecker`
* `MetadataChecker`
* `ResourceChecker`
* `ResourceMetadataChecker`

...?

stof replied:
I would say ResourceChecker

mpdude replied:
ResourceCheckerInterface, that is?

stof replied:
for the interface, yes

stof commented on src/Symfony/Component/Config/Tests/ConfigCacheTest.php:
```diff
@@ -12,29 +12,20 @@
     protected function tearDown()
     {
-        $files = array($this->cacheFile, $this->metaFile, $this->resourceFile);
+        $files = array($this->cacheFile, "{$this->cacheFile}.meta");
```
you should use concatenation rather than interpolation

stof commented on src/Symfony/Component/Config/ValidatorConfigCache.php:
```diff
@@ -0,0 +1,138 @@
+                }
+                if ($validator->isFresh($resource, $time)) {
+                    break; // no need to further check this resource
+                } else {
```
`` else`` is useless as you break the iteration in the if

stof commented:
as discussed in #15719, I think we should keep a new ResourceInterface for resources instead of using ``object`` in the MetadataValidator. And this new interface should be accepted in places using resources in Symfony

mpdude committed:
Tweaks as suggested on GH

mpdude commented:
Maybe I have a completely wrong picture of what the transition path for such changes must look like.

For example, the [MessageCatalogueInterface](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Translation/MessageCatalogueInterface.php#L171) is `@api` and contains an `addResource(ResourceInterface ...)` method. I don't see how we could reasonably switch to a new resource interface there? We must have a path (with `E_USER_DEPRECATED` messages if possible?) over 2.8, right?

This is why I am trying to clean up the non-`@api` `ResourceInterface` (see #15719). 

But if you can show me the right way to introduce a new interface for resources, I'd be glad to give it a try.

mpdude commented:
What about this plan:

* Add priorities so that new custom ResourceCheckers can be run first
* Provide a default ResourceChecker *with a very low priority* that will validate through `ResourceInterface::isFresh()` but triggers a deprecation warning.
* Add a new interface that inherits ResourceInterface and contains the `isFresh()` method. 
* Add a ResourceChecker for this interface with a slightly higher priority and *without* deprecation warning.
* Make FileResource et. al. implement the new interface.
* Deprecate ResourceInterface::isFresh() but keep that method in the new interface subclass.
* Also deprecate ResourceInterface::getResource() as per #15719 

We'll end up with ResourceInterface containing only __toString() in 3.0. 

I acknowledge that @stof is in favor of having the new interface extend \Serializable as well, but I am not sure if that is possible in a BC way *and* still think that technically it is not necessary (i. e. should not stop us from getting all the rest).

mpdude committed:
Implement the plan(TM)

* Add priorities so that new custom ResourceCheckers can be run first
* Provide a default ResourceChecker with a very low priority that will validate through ResourceInterface::isFresh() but triggers a deprecation warning.
* Add a new interface that extends ResourceInterface and contains the isFresh() method.
* Add a ResourceChecker for this interface with a slightly higher priority and without deprecation warning.
* Make FileResource et. al. implement the new interface.
* Deprecate ResourceInterface::isFresh() but keep that method in the new interface subclass.

mpdude commented:
I have implemented the above plan in #15738 and would like to close this PR in favor of it.

mpdude commented:
@bnw could you please close this?
